File: mascii2/src/test/java/com/kastkode/mascii2/musicelements/MetaInfoTests.java
package com.kastkode.mascii2.musicelements;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.List;

import org.junit.jupiter.api.Test;

import com.kastkode.mascii2.Mascii2ApplicationTests;

public class MetaInfoTests {

	@Test
	public void patchAssignment() {
		String minuet = "{patch: 5,7,9}\n\n"
				+ "g \n c \n a";
		List<Part> parts = Mascii2ApplicationTests.parse(minuet).getParts();
		assertEquals(3, parts.size());
		assertEquals(5 + "", parts.get(0).getMetaInfoChanges().get(0).rawValue);
		assertEquals(7 + "", parts.get(1).getMetaInfoChanges().get(0).rawValue);
		assertEquals(9 + "", parts.get(2).getMetaInfoChanges().get(0).rawValue);
	}	

	@Test
	public void commentsNBlankLines() {
		String minuet = "{\n -- comment here \n \n patch: 5,7,9}\n\n"
				+ " ------ another comment here ------ \n"
				+ "g \n c \n a";
		List<Part> parts = Mascii2ApplicationTests.parse(minuet).getParts();
		assertEquals(3, parts.size());
		assertEquals(5 + "", parts.get(0).getMetaInfoChanges().get(0).rawValue);
		assertEquals(7 + "", parts.get(1).getMetaInfoChanges().get(0).rawValue);
		assertEquals(9 + "", parts.get(2).getMetaInfoChanges().get(0).rawValue);
	}	
	
	@Test
	public void patchCountDoesntMatchPartCount() {

		String minuet = "{\n" + 
				"    tempo: 90\n" + 
				"    time:  4/4\n" + 
				"    key:   0\n" + 
				"    patch: 103\n" + 
				"}\n" + 
				"g \n" + 
				"C \n" + 
				"";
		
		List<Part> parts = Mascii2ApplicationTests.parse(minuet).getParts();
	}
}

File: mascii2-typescript/test/musicelements/MetaInfoTests.ts
import { test, describe } from 'node:test';
import assert from 'node:assert/strict';
import { SourceParser } from '../../src/SourceParser';
import { ParseResult } from '../../src/ParseResult';

function parse(source: string): ParseResult {
    const result = new SourceParser().generateFromString(source);
    const msgs = result.errors?.getMsgs() ?? [];
    for (const msg of msgs) console.error(msg);
    assert.equal(msgs.length, 0, `Expected no parse errors, got: ${msgs.join(', ')}`);
    return result;
}

describe('MetaInfoTests', () => {

    test('patchAssignment', () => {
        const minuet = '{patch: 5,7,9}\n\ng \n c \n a';
        const parts = parse(minuet).getParts() ?? [];
        assert.equal(3, parts.length);
        assert.equal('5', parts[0]!.getMetaInfoChanges()[0]!.getRawValue());
        assert.equal('7', parts[1]!.getMetaInfoChanges()[0]!.getRawValue());
        assert.equal('9', parts[2]!.getMetaInfoChanges()[0]!.getRawValue());
    });

    test('commentsNBlankLines', () => {
        const minuet = '{\n -- comment here \n \n patch: 5,7,9}\n\n'
            + ' ------ another comment here ------ \n'
            + 'g \n c \n a';
        const parts = parse(minuet).getParts() ?? [];
        assert.equal(3, parts.length);
        assert.equal('5', parts[0]!.getMetaInfoChanges()[0]!.getRawValue());
        assert.equal('7', parts[1]!.getMetaInfoChanges()[0]!.getRawValue());
        assert.equal('9', parts[2]!.getMetaInfoChanges()[0]!.getRawValue());
    });

    test('patchCountDoesntMatchPartCount', () => {
        const minuet = '{\n'
            + '    tempo: 90\n'
            + '    time:  4/4\n'
            + '    key:   0\n'
            + '    patch: 103\n'
            + '}\n'
            + 'g \n'
            + 'C \n';
        // should not throw
        parse(minuet).getParts();
    });

});
